Add Style.ShowHeaders to TableView

Nutzzz opened the pull request:
Very simple change to allow table headers to be hidden (necessary for e.g., using TableView as a single-dimension ListView split into multiple columns).  I'm also currently looking at methods of making such a use case easier (with options for horizontal or vertical item population and horizontal or vertical scrolling).

## Pull Request checklist:

- [X] I've named my PR in the form of "Fixes #issue. Terse description."
- [X] My code follows the [style guidelines of Terminal.Gui](https://github.com/gui-cs/Terminal.Gui/blob/develop/.editorconfig) - if you use Visual Studio, hit `CTRL-K-D` to automatically reformat your files before committing.
- [X] My code follows the [Terminal.Gui library design guidelines](https://github.com/gui-cs/Terminal.Gui/blob/develop/CONTRIBUTING.md)
- [X] I ran `dotnet test` before commit
- [X] I have made corresponding changes to the API documentation (using `///` style comments)
- [X] My changes generate no new warnings
- [X] I have checked my code and corrected any poor grammar or misspellings
- [X] I conducted basic QA to assure all features are working

tznind commented:
Thanks for looking into this use case.  It sounds exciting.

When new functionality is added to a View there are 2 things we need in addition to the code change:

- A scenario (or change to existing scenario) that demonstrates the feature.
- One or more tests that ensure the feature works

Could you please add a setting in `TableEditor` scenario for toggling this feature.  See `miHeaderOverline` and the other fields for how to do this.

Also please add some tests to `TableViewTests` that confirm the rendering is as expected in different situations.  See tests like `TableView_ExpandLastColumn_False_ExactBounds` as an example.

tznind commented on Terminal.Gui/Views/TableView/TableView.cs:
```diff
@@ -1772,6 +1775,11 @@ public class TableStyle {
 			public bool AlwaysShowHeaders { get; set; } = false;
 
+			/// <summary>
+			/// Hide header
```
Please elaborate on this comment e.g.:

> Gets or sets a flag indicating whether to suppress rendering headers of a <see cref="TableView"/>.  Defaults to false.

tznind commented on Terminal.Gui/Views/TableView/TableView.cs:
```diff
@@ -315,13 +315,16 @@ private int GetHeaderHeightIfAny ()
 		/// <returns></returns>
 		private int GetHeaderHeight ()
 		{
```
Could this be simplified by just adding the following at the top of the method instead?  Typically having less nesting is easier to read and maintain so would appreciate if you could make this change.

```csharp
if (Style.HideHeaders) {
    return 0;
}
```

tznind commented:
Bugs

- Sometimes the Back arrow will take you from "c:\somedir\" back to "c:\somedir".  These duplicates should not happen
- Tab Autocomplete has a case sensitivity issue where if you have a file system:

```csharp
/folder/item1
/folder/item2
```

And you enter a capital I (i.e. `"/folder/I"`) then it will correctly propose "item1" and tab completion will work BUT if youpress Down arrow the focus will move to table view (i.e. it will not cycle suggestions).

tznind commented on Terminal.Gui/Views/TableView/TableView.cs:
```diff
@@ -315,13 +315,16 @@ private int GetHeaderHeightIfAny ()
 		/// <returns></returns>
 		private int GetHeaderHeight ()
 		{
-			int heightRequired = 1;
```
Could this be simplified by just adding the following at the top of the method instead?  Typically having less nesting is easier to read and maintain so would appreciate if you could make this change.

```csharp
if (Style.HideHeaders) {
    return 0;
}
```

tig reviewed (approved):
I approve, but @tznind has final call.

Nutzzz commented:
@tznind Sorry, I decided to change this to allow the top border to be drawn even when HideHeaders is enabled.  Can you re-review?  I modified the TableEditor scenario and added Unit Tests.

tznind commented:
Looks good to me.  Thanks for taking the time to write the tests 🙂.

I'm on mobile atm so can't test but it looks perfect.  Happy for it to
merge.

On Mon, 10 Apr 2023, 18:21 Nutzzz, ***@***.***> wrote:

> @tznind <https://github.com/tznind> Sorry, I decided to change this to
> allow the top border to be drawn even when HideHeaders is enabled. Can you
> re-review? I modified the TableEditor scenario and added Unit Tests.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gui-cs/Terminal.Gui/pull/2514#issuecomment-1502084027>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHO3C5FYFF75HIGDVIHCKT3XAQ6SVANCNFSM6AAAAAAWW7OEIA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

tig commented on Terminal.Gui/Views/TableView/TableView.cs:
```diff
@@ -315,16 +316,16 @@ private int GetHeaderHeightIfAny ()
-			if (!Style.HideHeaders) {
-				heightRequired++;
+			if (Style.HideHeaders)
+				return 0;
```
Please use brackets for all nested statements. Even single lines. Thanks.

Nutzzz replied:
This was mooted, but thanks for the reminder.

tig replied:
> This was mooted, but thanks for the reminder.

Not sure what you mean by `mooted`. 

Before I can merge this, I need you to add brackets to all `if` statements, even if just a single line.

tig reviewed (commented):
@tznind needs to approve this.

Please see my comment.

tznind commented:
I am looking at #2543 and the style options.  I think I want to tweak this a little if that is ok.  All the other options begin with `Show` 

![image](https://user-images.githubusercontent.com/31306100/232181865-af5fcceb-a51a-485c-8363-c383610c45f7.png)

It makes me think this option should be `ShowHeaders` and default to `true`

I also think you should be able to mix and match both `ShowHorizontalHeaderOverline` and `ShowHorizontalHeaderUnderline` without caring about whether `ShowHeaders` is true.

I will submit a PR to your branch

tznind commented:
Here is the PR @Nutzzz  https://github.com/Nutzzz/Terminal.Gui/pull/1 

I think this simplifies the API.  Turning off headers does not change the other lines that are drawn.  Critically I think the underline is important as an option because it displays horizontal scroll indicators on that line so should be an option even without headers.

![image](https://user-images.githubusercontent.com/31306100/232182719-e6f1c0f9-555b-4deb-83ff-61f6034ec57e.png)
_TableEditor scenario with ShowHeaders changed to false_

Nutzzz commented:
Yeah, thanks. These changes work for me.

As far as the name I didn't like mixing Hide and Show either, but it was to avoid confusion  with AlwaysShowHeaders, which ought to have been called KeepHeadersOnscreen or something.

As for the underline, I didn't allow it just because it looks pretty weird. While the indicator is helpful, it seems like it could be moved somewhere... else? ...when the underline is disabled, since you're using up a whole line for the benefit of one or two characters.  A scrollbar would be better.  But you're right, let the developer figure out what they want to do.
